core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java

```diff
@@ -62,10 +65,20 @@
   public ProjectJoinTransposeRule(
       PushProjector.ExprCondition preserveExprCondition,
       RelBuilderFactory relFactory) {
-    super(
-        operand(Project.class,
-            operand(Join.class, any())),
-        relFactory, null);
+    this(
+        operand(LogicalProject.class,
+            operand(LogicalJoin.class, any())),
+            preserveExprCondition, relFactory);
+  }
+
+  /**
+   * Creates a ProjectJoinTransposeRule.
+   */
+  public ProjectJoinTransposeRule(
+      RelOptRuleOperand operand,
+      PushProjector.ExprCondition preserveExprCondition,
+      RelBuilderFactory relFactory) {
+    super(operand, relFactory, null);
```

Comment:
Do we really need a new constructor ?

Reply:
Just as @zabetak said "I think the way to go is to allow clients to specify their own Project, Join and RelBuilderFactory classes. I guess our default INSTANCE of this rule should be adapted to match only LogicalProject and LogicalJoin"
I think it's a good way, it gives more choice.

Reply:
I really mean that only one constructor is enough. You can construct INSTANCE by 
```
  public static final ProjectJoinTransposeRule INSTANCE =
      new ProjectJoinTransposeRule(
          operand(LogicalProject.class, operand(LogicalJoin.class, any())),
          expr -> !(expr instanceof RexOver),
          RelFactories.LOGICAL_BUILDER);

Reply:
The new constructor is more liberal than it should be; allowing to pass a RelOptRuleOperand means that the rule may match something else than Project and Join for which it was intended for. 

@jinxing64 Dropping a public constructor breaks backward compatibility so ideally it should remain as is (possibly deprecated). Other than that I agree with you that we can call directly the new constructor for creating the instance.

Reply:
Thanks for explanation @zabetak 
I got it now :)

Reply:
Thanks a lot for reviewing , I have addressed it.